rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,658 @@
+            // we can use any of the methods on it, or the traits that it requires (like `Reflect`)
+            dev_tool.toggle();
+        }
+    })
+}
+
+
+```
+
+Next, we want to be able to configure modal dev tools at run time.
+
+```rust
+#[derive(Event)]
+struct ConfigureDevTool{
+    tool_string: String,
+};
+
+fn configure_dev_tools(world: &mut World){
+    // Move the events out of the world, clearing them and avoiding a persistent borrow
+    let events = world.resource_mut::<Events<ToggleDevTool>>().drain();
+
+    // Use a resource scope to allow us to access both the dev tools registry and the rest of the world at the same time
+    world.resource_scope(|(world, dev_tools_registry: Mut<DevToolsRegistry>)|{
+        for event in events {
+            // Check the implementation details for information about how this works!
+            let result = dev_tools_registry.parse_and_insert_tool(event.tool_string);
+            if let Err(error) = result {
+                warn!(error);
+            }
+        }
+    })}
+```
+
+Finally, we want to be able to pass in a user supplied string, parse it into a dev command and then evaluate it on the world.
+
+```rust
+#[derive(Event)]
+struct DevCommandInput(String);
+
+fn parse_and_run_dev_commands(world: &mut World){
+    // Move the events out of the world, clearing them and avoiding a persistent borrow
+    let events = world.resource_mut::<Events<ToggleDevTool>>().drain();
+
+    // Use a resource scope to allow us to access both the dev tools registry and the rest of the world at the same time
+    world.resource_scope(|(world, dev_tools_registry: Mut<DevToolsRegistry>)|{
+        for event in events {
+            // This gives us access to the metadata needed to inspect the dev command and construct a new one.
+            let Some(dev_command_metadata) = dev_tools_registry.get_command_metadata_by_name(world, event.name) else {
+                warn!("No dev tool was found for {}). Did you forget to register it?", event.name);
+                continue;
+            };
+
+            // Create a concrete instance of our dev command from the supplied string.
+            let Ok(command) = dev_command_metadata.from_str(&event.0) else {
+                warn!("Could not parse the command from the supplied string");
+                continue;
+            }
+
+            // Now we can run the command directly on the `World` using `Command::apply`!
+            command.apply(world);
+        }
+    })
+}
+```
+
+While a number of other features could sensibly be added to this API (a `--help` flag, saving and loading config to disk, managing compatibility between dev tools),
+this MVP should be sufficient to prove out the viability of the core architecture.
+
+## Implementation strategy
+
+### What metadata do toolboxes need?
+
+We can start simple:
+
+- the type name of the dev tool
+- any doc strings on the dev tool struct
```

Comment:
It also seems to be suggested to be implemented, but is mentioned in passing in two parts of the text. Would you like to add a doc() method to ModalDevTool and DevCommand? Or is it intended to retrieve documentation via reflection, as mentioned in the comments before traits? I think the doc part is very unclear in rfc